internal/primitives/api/api.go

```diff
@@ -0,0 +1,84 @@
+// Copyright 2025 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package api
+
+import (
+	"github.com/ChainSafe/gossamer/internal/primitives/core"
+	"github.com/ChainSafe/gossamer/internal/primitives/runtime"
+	statemachine "github.com/ChainSafe/gossamer/internal/primitives/state-machine"
+	"github.com/ChainSafe/gossamer/internal/primitives/state-machine/overlayedchanges"
+	ptrie "github.com/ChainSafe/gossamer/internal/primitives/trie"
+	"github.com/ChainSafe/gossamer/internal/primitives/trie/recorder"
+)
+
+type ProvideRuntimeApi[
+	N runtime.Number,
+	E runtime.Extrinsic,
+	H runtime.Hash,
+	Hasher runtime.Hasher[H],
+	Backend statemachine.Backend[H, Hasher],
+	Result any,
+	Api ApiExt[N, E, H, Hasher, Backend, Result],
+] interface {
+	// Returns the runtime api.
+	// The returned instance will keep track of modifications to the storage. Any successful
+	// call to an api function, will `commit` its changes to an internal buffer. Otherwise,
+	// the modifications will be `discarded`. The modifications will not be applied to the
+	// storage, even on a `commit`.
+	RuntimeApi() Api
+}
+
+type ConstructRuntimeApi[
+	N runtime.Number,
+	E runtime.Extrinsic,
+	H runtime.Hash,
+	Hasher runtime.Hasher[H],
+	Backend statemachine.Backend[H, Hasher],
+	Result any,
+	RuntimeApi ApiExt[N, E, H, Hasher, Backend, Result],
+] interface {
+	// Construct an instance of the runtime api.
+	ConstructRuntimeApi() RuntimeApi
+}
```

Comment:
Don't think you need all these generic params when you can specify what you'd like to return.
```suggestion
type ConstructRuntimeAPI[RuntimeAPI any] interface {
	// Construct an instance of the runtime api.
	ConstructRuntimeAPI() RuntimeAPI
}
```

Reply:
I think the same reply for `ProvideRuntimeAPI` applies here https://github.com/ChainSafe/gossamer/pull/4647#discussion_r2061627638

Reply:
simplified